internal/encryption/encryption.go

```diff
@@ -0,0 +1,341 @@
+package encryption
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/gohcl"
+	"github.com/opentofu/opentofu/internal/varhcl"
+	"github.com/zclconf/go-cty/cty"
+)
+
+// Encryption contains the methods for obtaining a State or Plan correctly configured for a specific
+// purpose. If no encryption configuration is present, it returns a passthru method that doesn't do anything.
+type Encryption interface {
+	StateFile() State
+	PlanFile() Plan
+	Backend() State
+	RemoteState(string) State
+}
+
+type encryption struct {
+	// Inputs
+	cfg *Config
+	reg Registry
+
+	// Used to evaluate hcl expressions
+	ctx *hcl.EvalContext
+	// Used to build EvalContext (and related mappings)
+	keyValues    map[string]map[string]cty.Value
+	methodValues map[string]map[string]cty.Value
+	methods      map[string]Method
+
+	stateFile     State
+	planFile      Plan
+	backend       State
+	remoteDefault State
+	remote        map[string]State
+}
+
+func New(reg Registry, cfg *Config) (Encryption, hcl.Diagnostics) {
+	var diags hcl.Diagnostics
+
+	enc := &encryption{
+		cfg: cfg,
+		reg: reg,
+
+		ctx: &hcl.EvalContext{
+			Variables: map[string]cty.Value{},
+		},
+		keyValues:    make(map[string]map[string]cty.Value),
+		methodValues: make(map[string]map[string]cty.Value),
+		methods:      make(map[string]Method),
+
+		remote: make(map[string]State),
+	}
+
+	// TODO handle cfg == nil
+
+	diags = append(diags, enc.setupKeyProviders()...)
+	if diags.HasErrors() {
+		return nil, diags
+	}
+	diags = append(diags, enc.setupMethods()...)
+	if diags.HasErrors() {
+		return nil, diags
+	}
+	diags = append(diags, enc.setupTargets()...)
+	if diags.HasErrors() {
+		return nil, diags
+	}
+
+	return enc, diags
+}
+
+func (enc encryption) setupKeyProviders() hcl.Diagnostics {
```

Comment:
constructor returns a *encryption, but methods are implemented using a non-pointer receiver.

This could lead to problems if a field gets added to encryption that should not be copied.

For example, do we need a mutex in encryption to synchronize access to all those maps, so the feature works consistently with parallel tests?